src/bench.rs

```diff
@@ -1,133 +1,361 @@
+    }
+}
+
+/// Benchmark results.
+pub struct Results {
+    /// Sorted samples for calculations like median.
+    sorted_samples: Vec<usize>,
+
+    /// Samples ordered chronologically.
+    samples: Vec<usize>,
+
+    /// Number of bytes in one sample of the benchmark.
+    bench_size: usize,
+
+    /// Name of the benchmark.
+    name: String,
 }
 
-pub fn unicode_random_write<W: Write>(ctx: &mut Context<W>, options: &Options) -> Result<usize> {
-    let mut written = 0;
-    let mut rng = rand::thread_rng();
-    let h = options.height;
-    let w = options.width;
+#[allow(unused)]
+impl Results {
+    pub fn new(name: String, bench_size: usize, mut samples: Vec<usize>) -> Self {
+        // Assure the vector is never empty to simplify the math.
+        if samples.is_empty() {
+            samples.push(0);
+        }
+
+        let mut sorted_samples = samples.clone();
+        sorted_samples.sort_unstable();
+
+        Results { sorted_samples, samples, bench_size, name }
+    }
+
+    /// Name of the benchmark.
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    /// All recorded samples.
+    pub fn samples(&self) -> &[usize] {
+        self.samples.as_slice()
+    }
+
+    /// Size of the benchmark in bytes.
+    pub fn bench_size(&self) -> usize {
+        self.bench_size
+    }
 
-    while written < options.bytes {
-        let (l, c) = (rng.gen_range(0, h), rng.gen_range(0, w));
+    /// Number of samples taken.
+    pub fn sample_count(&self) -> usize {
+        self.samples.len()
+    }
+
+    /// Fastest benchmark sample in milliseconds.
+    pub fn min(&self) -> usize {
+        self.samples.iter().min().copied().unwrap_or_default()
+    }
 
-        written += ctx.cup(l, c)?;
-        if options.colorize {
-            written += ctx.setaf(rng.gen_range(0, 8))?;
+    /// Slowest benchmark sample in milliseconds.
+    pub fn max(&self) -> usize {
+        self.samples.iter().max().copied().unwrap_or_default()
+    }
+
+    /// Mean execution time per sample in milliseconds.
+    pub fn mean(&self) -> f64 {
+        self.samples.iter().sum::<usize>() as f64 / self.samples.len() as f64
+    }
+
+    /// Median execution time per sample in milliseconds.
+    pub fn median(&self) -> f64 {
+        let len = self.samples.len();
+        (self.sorted_samples[(len - 1) / 2] as f64 + self.sorted_samples[len / 2] as f64) / 2.
+    }
+
+    /// Variance of the execution time in milliseconds.
+    pub fn variance(&self) -> f64 {
+        if self.samples.len() < 2 {
+            return 0.;
         }
 
-        let unicode_value = rng.gen_range(0, u16::max_value());
-        let unicode = String::from_utf16_lossy(&[unicode_value]).into_bytes();
-        ctx.write_all(&unicode)?;
-        written += unicode.len();
+        let last_index = (self.samples.len() - 1) as f64;
+        let mean = self.mean() as f64;
+        self.samples.iter().map(|&s| f64::powi(s as f64 - mean, 2)).sum::<f64>() / last_index
```

Comment:
I've extracted it mainly for formatting reasons, but I don't see how `last_index` would be confusing? What do you think could be misunderstood here?

Reply:
Well, certain math formulas are using `i=1,n` notation for iteration, and the last index here is `n`, however you want `n - 1`. And indices are usually starting from 1 in math, so it confused me a bit, since the technical last_index in a `sum` is `n`, but you want `n - 1`. Does it make sense or it's just me being a bit weird?

Reply:
Oh, and I'm fine with extracting, I just don't like the name, however I can't really say how it should be called, so I decided to suggest doing `-1` in place(it'll look more like formula you've used, so easier to understand).

Reply:
I think I actually confused myself here with the index starting at 1. You're correct that last_index is not the right name.